### PR TITLE
Add online recipe to aishell1

### PR DESCRIPTION
**Pull request by naxingyu**
This PR is for addressing #2801 . The config of online pitch is still being evaluated.
The online recipe will be named _2a in the tuning folder, while the running scripts will still be linked to the offline recipes. Online results will be provided.

### CODE REVIEW

**Comment by naxingyu** [comment]
The chain model results, including CER and average online RTF, are presented below. 
Which config should I commit? @danpovey 

 chain         |right-context=75 | right-context=25
------------|-------------------|-------------------
test            |    7.44%   | 7.53% 
+online      |   7.46% (RTF=1.044)   |  7.55% (RTF=1.026)
+per-utt   | 7.90%  | 8.03%

**Comment by danpovey** [comment]
Fantastic!
BTW, the right-context isn't expected to impact the RTF, it affects the
latency though (0.25 vs. 0.75 secnods).
I think the right-context=25 setup makes more sense to commit, given that
the difference in WER is so small.

**Comment by danpovey** [comment]
BTW, let me know how this compares with the previous, non-online-pitch
results.

**Comment by naxingyu** [comment]
Comparison with the previous setup.

non-online-pitch | 7.46%
------------------|----------
online-pitch ctx=25 | 7.53%
online-pitch +online decoding | 7.55%

**Comment by danpovey** [comment]
Cool-- thanks a lot.
Can you remind me how much pitch helps in relative %WER (generally, not
necessarily in this specific setup)?
Need to know for some future plans.

**Comment by naxingyu** [comment]
Adding pitch itself normally reduce WER by relatively 10%~15%. Adding pitch related segmental features, such as NCCF, can further reduce the WER, but not much. The reduction could be larger in tonal language SR if the phone set is tonal.

**Comment by danpovey** [comment]
That's interesting.  Can you add details here?  What do you mean by
"segmental", and are you talking about, say, the chosen or peak NCCF value,
or the whole array of NCCFs for different lags?  And are you are aware of
any prior work on attempting to incorporate low-latency pitch-related
features for ASR?

**Comment by naxingyu** [comment]
This PR is ready for merge.

**Comment by xiaosdawn** [comment]
Hi，
Recently, I have trained my chain model using my own data. 
I just used MFCC for network input, when training. The details of my script as follows:

`
if [ $stage -le 12 ]; then
  echo "$0: creating lang directory $lang with chain-type topology"
  # Create a version of the lang/ directory that has one state per phone in the
  # topo file. [note, it really has two states.. the first one is only repeated
  # once, the second one has zero or more repeats.]
  if [ -d $lang ]; then
    if [ $lang/L.fst -nt data/lang/L.fst ]; then
      echo "$0: $lang already exists, not overwriting it; continuing"
    else
      echo "$0: $lang already exists and seems to be older than data/lang..."
      echo " ... not sure what to do.  Exiting."
      exit 1;
    fi
  else
    cp -r data/lang $lang
    silphonelist=$(cat $lang/phones/silence.csl) || exit 1;
    nonsilphonelist=$(cat $lang/phones/nonsilence.csl) || exit 1;
    # Use our special topology... note that later on may have to tune this
    # topology.
    steps/nnet3/chain/gen_topo.py $nonsilphonelist $silphonelist >$lang/topo
  fi
fi

if [ $stage -le 13 ]; then
  # Get the alignments as lattices (gives the LF-MMI training more freedom).
  # use the same num-jobs as the alignments
  nj=$(cat $ali_dir/num_jobs) || exit 1;
  steps/align_fmllr_lats.sh --nj $nj --cmd "$train_cmd" ${lores_train_data_dir} \
     data/lang $gmm_dir $lat_dir
  rm $lat_dir/fsts.*.gz # save space
fi

if [ $stage -le 14 ]; then
  # Build a tree using our new topology. This is the critically different
  # step compared with other recipes.
  if [ -f $treedir/final.mdl ]; then
     echo "$0: $treedir/final.mdl already exists, refusing to overwrite it."
     exit 1;
  fi
  steps/nnet3/chain/build_tree.sh --frame-subsampling-factor 3 \
      --leftmost-questions-truncate $leftmost_questions_truncate \
      --context-opts "--context-width=2 --central-position=1" \
      --cmd "$train_cmd" 7000 ${lores_train_data_dir} $lang $ali_dir $treedir
fi

if [ $stage -le 15 ]; then
  echo "$0: creating neural net configs using the xconfig parser";
  num_targets=$(tree-info $treedir/tree |grep num-pdfs|awk '{print $2}')
  learning_rate_factor=$(echo "print 0.5/$xent_regularize" | python)
  opts="l2-regularize=0.01"
  output_opts="l2-regularize=0.0025"

  mkdir -p $dir/configs
  cat <<EOF > $dir/configs/network.xconfig
  
  input dim=40 name=input

  # please note that it is important to have input layer with the name=input
  # as the layer immediately preceding the fixed-affine-layer to enable
  # the use of short notation for the descriptor
  fixed-affine-layer name=lda input=Append(-2,-1,0,1,2) affine-transform-file=$dir/configs/lda.mat

  # the first splicing is moved before the lda layer, so no splicing here
  relu-batchnorm-layer name=tdnn1 dim=1024
  relu-batchnorm-layer name=tdnn2 input=Append(-2,-1,0,1,2) dim=1024
  relu-batchnorm-layer name=tdnn3 input=Append(-1,2) dim=1024
  relu-batchnorm-layer name=tdnn4 input=Append(-2,1) dim=1024
  relu-batchnorm-layer name=tdnn5 input=Append(-3,3) dim=1024
  relu-batchnorm-layer name=tdnn6 input=Append(-5,1) dim=1024
  relu-batchnorm-layer name=tdnn7 input=Append(0) dim=1024

  ## adding the layers for chain branch
  relu-batchnorm-layer name=prefinal-chain input=tdnn7 dim=1024 target-rms=0.5
  output-layer name=output include-log-softmax=false dim=$num_targets max-change=1.5

  # adding the layers for xent branch
  # This block prints the configs for a separate output that will be
  # trained with a cross-entropy objective in the 'chain' models... this
  # has the effect of regularizing the hidden parts of the model.  we use
  # 0.5 / args.xent_regularize as the learning rate factor- the factor of
  # 0.5 / args.xent_regularize is suitable as it means the xent
  # final-layer learns at a rate independent of the regularization
  # constant; and the 0.5 was tuned so as to make the relative progress
  # similar in the xent and regular final layers.
  relu-batchnorm-layer name=prefinal-xent input=tdnn7 dim=1024 target-rms=0.5
  output-layer name=output-xent dim=$num_targets learning-rate-factor=$learning_rate_factor max-change=1.5

EOF
  steps/nnet3/xconfig_to_configs.py --xconfig-file $dir/configs/network.xconfig --config-dir $dir/configs/
fi

if [ $stage -le 16 ]; then
  steps/nnet3/chain/train.py --stage $train_stage \
    --cmd "$decode_cmd" \
    --feat.cmvn-opts "--norm-means=true --norm-vars=false" \
    --chain.xent-regularize $xent_regularize \
    --chain.leaky-hmm-coefficient 0.1 \
    --chain.l2-regularize 0.00005 \
    --chain.apply-deriv-weights false \
    --chain.lm-opts="--num-extra-lm-states=2000" \
    --egs.dir "$common_egs_dir" \
    --egs.stage $get_egs_stage \
    --egs.opts "--frames-overlap-per-eg 0" \
    --egs.chunk-width $frames_per_eg \
    --trainer.srand=$srand \
    --trainer.num-chunk-per-minibatch $minibatch_size \
    --trainer.frames-per-iter 1500000 \
    --trainer.num-epochs $num_epochs \
    --trainer.optimization.num-jobs-initial $num_jobs_initial \
    --trainer.optimization.num-jobs-final $num_jobs_final \
    --trainer.optimization.initial-effective-lrate $initial_effective_lrate \
    --trainer.optimization.final-effective-lrate $final_effective_lrate \
    --trainer.max-param-change $max_param_change \
    --cleanup.remove-egs $remove_egs \
    --feat-dir $train_data_dir \
    --tree-dir $treedir \
    --lat-dir $lat_dir \
    --dir $dir  || exit 1;
fi

if [ $stage -le 17 ]; then
  # Note: it might appear that this $lang directory is mismatched, and it is as
  # far as the 'topo' is concerned, but this script doesn't read the 'topo' from
  # the lang directory.
  utils/lang/check_phones_compatible.sh data/lang_test/phones.txt $lang/phones.txt
  utils/mkgraph.sh --self-loop-scale 1.0 data/lang_test $dir $dir/graph
fi

if [ $stage -le 18 ];then
  ## offline
  if $decode_offline ; then
    for decode_set in test; do
       (
       steps/nnet3/decode.sh --acwt 1.0 --post-decode-acwt 10.0 \
           --nj 4 --cmd "$decode_cmd" \
           $graph_dir $test_data_dir/${decode_set}_hires \
           $dir/decode_${test_data}_${decode_set} || exit 1;
       ) || touch $dir/.error &
    done
    wait
    if [ -f $dir/.error ]; then
      echo "$0: something went wrong in decoding"
      exit 1
    fi
  fi

  ## online
  if $decode_online ; then
     steps/online/nnet3/prepare_online_decoding.sh --mfcc-config conf/mfcc_hires.conf \
          $lang "$dir" ${dir}_online || exit 1;
          #$lang "$dir" ${dir}_online || exit 1;

     dir=${dir}_online
     for test_set in test; do
       steps/online/nnet3/decode.sh --acwt 1.0 --post-decode-acwt 10.0 \
           --nj 10 --cmd "$decode_cmd" \
           --config conf/decode.config \
           $graph_dir $test_data_dir/${test_set}_hires $dir/decode_${test_set} || exit 1;
    done
  fi
fi
`

The offline result is right and the WER is 6.65%. But the online result is null, which means every utterence is null, the WER is 100%. Some configuration must be mismatch, but I don't know where it is.
Is there anything I missed ? Looking forward to replay. Thx.

**Comment by danpovey** [comment]
maybe you used a different MFCC config to produce the MFCC features?

**Comment by danpovey** [comment]
also you should be asking on kaldi-help, not here.

**Comment by xiaosdawn** [comment]
OK，sorry for that，thank you，Daniel.发自我的iPhone------------------ Original ------------------From: Daniel Povey  <notifications@github.com>Date: Thu,Jan 17,2019 1:16 PMTo: kaldi-asr/kaldi <kaldi@noreply.github.com>Cc: xiaosdawn <xiaos_dawn@163.com>, Comment <comment@noreply.github.com>Subject: Re: [kaldi-asr/kaldi] Add online recipe to aishell1 (#2829)also you should be asking on kaldi-help, not here.

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/kaldi-asr/kaldi","title":"kaldi-asr/kaldi","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/kaldi-asr/kaldi"}},"updates":{"snippets":[{"icon":"PERSON","message":"@danpovey in #2829: also you should be asking on kaldi-help, not here."}],"action":{"name":"View Pull Request","url":"https://github.com/kaldi-asr/kaldi/pull/2829#issuecomment-455045492"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/kaldi-asr/kaldi/pull/2829#issuecomment-455045492",
"url": "https://github.com/kaldi-asr/kaldi/pull/2829#issuecomment-455045492",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]
